pageserver/src/tenant/remote_timeline_client/download.rs

```diff
@@ -73,33 +74,129 @@ pub async fn download_layer_file<'a>(
+
+                        // Tokio doc here: https://docs.rs/tokio/1.17.0/tokio/fs/struct.File.html states that:
+                        // A file will not be closed immediately when it goes out of scope if there are any IO operations
+                        // that have not yet completed. To ensure that a file is closed immediately when it is dropped,
+                        // you should call flush before dropping it.
+                        //
+                        // From the tokio code I see that it waits for pending operations to complete. There shouldt be any because
+                        // we assume that `destination_file` file is fully written. I e there is no pending .write(...).await operations.
+                        // But for additional safety lets check/wait for any pending operations.
+                        destination_file
+                            .flush()
+                            .await
+                            .with_context(|| format!("flush source file at {temp_file_path}"))
+                            .map_err(DownloadError::Other)?;
+
+                        // not using sync_data because it can lose file size update
+                        destination_file
+                            .sync_all()
+                            .await
+                            .with_context(|| {
+                                format!("failed to fsync source file at {temp_file_path}")
+                            })
+                            .map_err(DownloadError::Other)?;
+
+                        Ok(bytes_amount)
+                    }
+                    .await
+                }
+                #[cfg(target_os = "linux")]
+                crate::virtual_file::io_engine::IoEngine::TokioEpollUring => {
+                    async {
+                        let destination_file = VirtualFile::create(&temp_file_path)
+                            .await
+                            .with_context(|| {
+                                format!("create a destination file for layer '{temp_file_path}'")
+                            })
+                            .map_err(DownloadError::Other)?;
+
+                        let mut download = storage.download(&remote_path, cancel).await?;
+
+                        // TODO: use vectored write (writev) once supported by tokio-epoll-uring.
+                        // There's chunks_vectored() on the stream.
+                        let (bytes_amount, destination_file) = async {
+                            let size_tracking = size_tracking_writer::Writer::new(destination_file);
+                            let mut buffered = owned_buffers_io::write::BufferedWriter::<
+                                { super::BUFFER_SIZE },
+                                _,
+                            >::new(size_tracking);
+                            while let Some(res) =
+                                futures::StreamExt::next(&mut download.download_stream).await
```

Comment:
How come this does not need pinning?

Reply:
Aah yes because it's boxpinned already, that's why. I was wondering if this was an older commit from before I lifted the Unpin requirement.